### PR TITLE
Add 3rd party Oauth support, webhooks, TargetID

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary

Uses https://github.com/mattermost/mattermost-plugin-apps/pull/114

Uses the proxy app webhook implementation. (https://github.com/mattermost/mattermost-plugin-apps/pull/114)
* Apps can now create the webhook path by adding to the webhook URL extracted from the proxy app with "/{app_id}/{path}"

README
* Update instructions to include new method for creating a Target in Zendesk
* Correct the links for redirect because now pointing to mm-site-url, not the node server hosting the zendesk app  

Add Target Creation Automation
  * Add `/zendesk setup-target` slash command
  * Remove Target ID from config modal
  * create target through the `setup-target` slash command
  * update README
    * remove all the target creation instructions and replace with new
      instructions

`clients/mattermost_proxy_plugin.ts`
* based off of the mmclient in proxy app.  
* used for 
	* KV operations (Set, Get, Delete)
	* Storing Oauth2App
	* Storing Oauth2User token

`bindings`
* configure custom call to include 
  * acting_user to get roles 
  * oauth2_app - to see if the oauth has been configured
  * oauth2_user - to see if the user is connected

Using Oauth2 with latest proxy integration
* client_id & client_secret are stored by the proxy app
* Values are now saved via API and retrieved via expanded oauth2

#### Fixes for Subscriptions (Discovered while testing with Dylan)
* Add instance to the trigger name. The search needs to be specific to the instance so multiple instances can use one ZD account.  

#### Ticket Link
- [x] https://mattermost.atlassian.net/browse/MM-34275 - Use 3rd party Oauth2 support PR
- [x] https://mattermost.atlassian.net/browse/MM-34373 - Remove hardcoded TargetID value from code
- [x] https://mattermost.atlassian.net/browse/MM-33938 - README - add screenshot of ticket create post
- [x] https://mattermost.atlassian.net/browse/MM-33940 - verify and use context.admin_access_token
- [x] https://mattermost.atlassian.net/browse/MM-33448 - Bindings - use context to determine isadmin
- [x] https://mattermost.atlassian.net/browse/MM-33749 - Webhooks - need access to an acting_user_id for creating posts
- [x] https://mattermost.atlassian.net/browse/MM-34721 - Zendesk App: Only sysadmin should be allowed to type Zendesk configure command
- [x] https://mattermost.atlassian.net/browse/MM-34722 - Zendesk App: Zendesk Client Secret should be masked and displayed in UI after it is saved once.
- [x] https://mattermost.atlassian.net/browse/MM-34145 - MM link in zendesk ticket is wrong
- [x] https://mattermost.atlassian.net/browse/MM-33750 - Automate Zendesk TargetID generation
- [ ] https://mattermost.atlassian.net/browse/MM-33703 - Update README - slash commands and screenshots images

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,31 +11,45 @@ mm-plugin-apps: `master` @831d203
...
-    1. `make run` - (in a separate shell) start the node server
+##### 1. Create Zendesk Oauth Client for Mattermost (in Zendesk)
+
+1. `Zendesk` > `Admin` > `API` > `OAuth Clients`
+1. `Add OAuth Client`
+    1. `Client Name`: (Example `Mattermost Zendesk App`)
+    1. `Description`: `Connect your Zendesk account to Mattermost`
+    1. `Redirect URLs`: `https://<your-zendesk-app-host>/oauth2/complete`
+        1. Ex. `http://localhost:4000` - Development
+        1. Ex. `https://mytest.ngrok.io` - Exposed for development
```
Comment (hanzei) [inline README.md:22]: Is that correct? Does that path have to contain the Apps plugin id?

Comment (jfrerich) [reply README.md]: @hanzei, with the latest Oauth2 changes, the redirect URL now points to the `mattermost_site_url`.  I've updated the README.

**Comment by jfrerich** [comment]
@dylan, this PR get's the app back in to a fully testable state.

**Comment by jfrerich** [comment]
@DHaussermann, I just pushed a rather significant change for replacing the Zendesk Target Creation.

* Added `/zendesk setup-target` slash command
  * creates target in zendesk 
  * running multiple times doesn't invalidate any triggers, but only needs to be run one time.
  * no longer requires user to create manually
  * Title is now created by the app
    * `Mattermost Subcriptions <mattermost-site-url>`
* Remove Target ID from config modal (now automated)
* README
  * remove all the target creation instructions and replace them with new
instructions

**Comment by jfrerich** [comment]
@hanzei, @levb, @mickmister, @catalintomai.  I'm sorry this PR is growing beyond the original scope.  I've got @DHaussermann working with this branch and solving numerous items we have found and would like to keep adding to this branch instead of creating numerous PRs.  It is ready for review, but will keep evolving.  If we get a good mergeable state, I'll merge and then create a new PR.

File: src/app/triggerFromForm.ts [hunk 2]
```diff
@@ -11,11 +11,13 @@ interface TriggerFromFrom {
 export class TriggerFromFormImpl implements TriggerFromFrom {
     values: AppFormValues;
     context: AppContext;
+    targetID: string;
     trigger: any
 
-    constructor(call: AppCall) {
+    constructor(call: AppCall, targetID: strinb) {
```
Comment (catalintomai) [inline src/app/triggerFromForm.ts:17]: s/strinb/string

File: src/bindings/channel_header.ts [hunk 3]
```diff
@@ -1,20 +1,20 @@
 import {AppBinding, AppContext} from 'mattermost-redux/types/apps';
 import {AppExpandLevels} from 'mattermost-redux/constants/apps';
 
-import {getStaticURL, Routes, newChannelHeaderBindings} from '../utils';
+import {getStaticURL, Routes, newChannelHeaderBindings, isConfigured, isConnected, isUserSystemAdmin} from '../utils';
 import {CommandLocations, ZendeskIcon} from '../utils/constants';
 
 // getChannelHeaderBindings returns the users command bindings
```
Comment (catalintomai) [inline src/bindings/channel_header.ts:7]: s/command/channel header.

File: src/bindings/channel_header.ts [hunk 4]
```diff
@@ -1,20 +1,20 @@
...
-export const getChannelHeaderBindings = (context: AppContext, configured: boolean, connected: boolean, sysadmin: boolean): AppBinding => {
+export const getChannelHeaderBindings = (context: AppContext): AppBinding => {
     const bindings: AppBinding[] = [];
+    const isSysadmin = isUserSystemAdmin(context);
 
     // only show configuration option if admin has not configured the plugin
-    if (!configured && sysadmin) {
+    if (!isConfigured(context) && isSysadmin) {
         bindings.push(channelHeaderConfig(context));
         return newChannelHeaderBindings(bindings);
```
Comment (catalintomai) [inline src/bindings/channel_header.ts:15]: could use an `else if` and remove this `return new` statement (and reuse the final one)

File: src/restapi/fIncomingWebhooks.ts [hunk 5]
```diff
@@ -11,31 +10,37 @@ import {newZDClient, newMMClient} from '../clients';
...
-        acting_user_id: 'rgixs6uimp88tq8x8w3yxu3oqe',
-    };
-    const zdClient = await newZDClient(fakeBotContext);
+    const ticketID = values[TriggerFields.TicketIDKey];
+    const channelID = values[TriggerFields.ChannelIDKey];
+
+    const config = await newConfigStore(context).getValues();
+    const zdHost = config.zd_node_host;
+
+    const connectedUser = config.zd_oauth_access_token;
```
Comment (catalintomai) [inline src/restapi/fIncomingWebhooks.ts:22]: technically a token, not a user.

File: src/app/triggerFromForm.ts (hunk 2 above)
Comment (jfrerich) [reply src/app/triggerFromForm.ts]: done.

File: src/restapi/fIncomingWebhooks.ts (hunk 5 above)
Comment (jfrerich) [reply src/restapi/fIncomingWebhooks.ts:22]: done.

File: src/bindings/channel_header.ts (hunk 3 above)
Comment (jfrerich) [reply src/bindings/channel_header.ts:7]: done.  I renamed CommandLocations to Locations this is agnostic.  This was left over from when I initially created the app with only command locations

File: src/bindings/channel_header.ts (hunk 4 above)
Comment (jfrerich) [reply src/bindings/channel_header.ts]: done.

File: src/clients/mattermost_proxy_plugin.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        this.url = baseURL;
+    }
+    async kvSet(key: string, value: any): Promise<void> {
+        const url = this.url + this.kvPath(key);
+        try {
+            this.doAPIPost(url, value);
+        } catch (e) {
+            console.log('e', e);
+        }
+    }
```
Comment (mickmister) [inline src/clients/mattermost_proxy_plugin.ts:35]: If we let the `doAPIPost` call resolve its promise before exiting this function, we can let callers of `kvSet` wait for the execution to be done by using `await kvSet()`. This also lets us be able to bubble an error from `doAPIPost` if necessary.

```suggestion
    async kvSet(key: string, value: any): Promise<void> {
        const url = this.url + this.kvPath(key);
        return this.doAPIPost(url, value);
    }
```

If you want to sanitize the error or change it, you can do that in a catch block and rethrow an error:

```go
try {
    return this.doAPIPost(url, value);
} catch (e) {
    throw new Error('setKv failed: ' + e.message);
}
```

Comment (jfrerich) [reply src/clients/mattermost_proxy_plugin.ts]: Done.

**Comment by jfrerich** [comment]
@catalintomai, @mickmister can you both please re-review? I would really like to get this merged today if possible

**Review by levb — APPROVED** [review]
It's a good improvement, I'd like to get this merged and iterate from there.

File: src/bindings/slash_commands.ts [hunk 7]
```diff
@@ -57,33 +66,78 @@ const cmdDisconnect = (context: AppContext): AppBinding => {
...
+        label: 'setup-target',
+        description: 'Setup Zendesk Target',
+        icon: getStaticURL(context, ZendeskIcon),
+        form: {},
+        call: {
+            path: Routes.App.BindingPathTargetCreate,
+            expand: {
+                app: AppExpandLevels.EXPAND_ALL,
+
+                // acting_user: AppExpandLevels.EXPAND_ALL,
```
Comment (hanzei) [inline src/bindings/slash_commands.ts:131]: Remove?

File: src/clients/mattermost_proxy_plugin.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
+import fetch from 'node-fetch';
+
+import {Oauth2App} from 'src/utils';
+
+import {Routes, AppsPluginName, PathAPI, PathKV} from '../utils/constants';
+import {tryPromiseWithMessage} from '../utils/utils';
+import {getManifest} from '../manifest';
+
+export interface ProxyClient {
```
Comment (hanzei) [inline src/clients/mattermost_proxy_plugin.ts:9]: 1/5 
```suggestion
export interface AppsClient {
```

File: src/clients/mattermost_proxy_plugin.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    kvSet(key: string, value: any): Promise<void>;
+    kvGet(key: string): Promise<any>;
+    kvDelete(key: string): void;
+    storeOauth2App(id: string, secret: string): void;
+    storeOauth2User(token: string): void
```
Comment (hanzei) [inline src/clients/mattermost_proxy_plugin.ts:14]: Should these be Promises?

File: src/clients/mattermost_proxy_plugin.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    storeOauth2App(id: string, secret: string): Promise<void> {
+        const url = this.url + this.oauth2AppPath();
+        const data: Oauth2App = {
+            client_id: id,
+            client_secret: secret,
+        };
+        return tryPromiseWithMessage(this.doAPIPost(url, data), 'storeOauth2App failed');
+    }
+
+    storeOauth2User(token: any): Promise<void> {
```
Comment (hanzei) [inline src/clients/mattermost_proxy_plugin.ts:53]: Can this be a `string`?

**Review by hanzei — COMMENTED** [review]
A few comments below

File: src/utils/constants.ts [hunk 11]
```diff
@@ -5,9 +5,16 @@ export const FormTextAreaMaxLength = 1024;
 
 import {AppSelectOption} from 'mattermost-redux/types/apps';
 
+export const AppsPluginName = 'com.mattermost.apps';
+export const PathAPI = '/api/v1';
+export const PathProxyAPI = `/plugins/${AppsPluginName}/${PathAPI}`;
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/utils/constants.ts:10]: yes, unused and removed.

File: src/bindings/slash_commands.ts (hunk 7 above)
Comment (jfrerich) [reply src/bindings/slash_commands.ts:131]: yep. removed.

File: src/clients/mattermost_proxy_plugin.ts (hunk 10 above)
Comment (jfrerich) [reply src/clients/mattermost_proxy_plugin.ts:53]: yes.  Done.

File: src/clients/mattermost_proxy_plugin.ts (hunk 8 above)
Comment (jfrerich) [reply src/clients/mattermost_proxy_plugin.ts]: Done.

File: src/clients/mattermost_proxy_plugin.ts (hunk 9 above)
Comment (jfrerich) [reply src/clients/mattermost_proxy_plugin.ts]: yes. Done!

File: src/manifest.json [hunk 12]
```diff
@@ -8,7 +8,11 @@
     "root_url": "",
     "requested_permissions": [
         "act_as_user",
-        "act_as_bot"
+        "act_as_bot",
+        "act_as_admin",
+        "act_as_sysadmin",
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/manifest.json:14]: Negative.  It doesn't, but I did also find it in https://github.com/mattermost/mattermost-plugin-apps/blob/0919e460b794fa57911745ccbc9d3e78b9306a4e/assets/example-aws-hello.json#L12

**Comment by jfrerich** [comment]
@hanzei did you have any other issues with the PR? Are you OK with approving for merge?

**Review by hanzei — APPROVED** [review]
LGTM

File: src/restapi/fConnect.ts [hunk 13]
```diff
@@ -23,5 +34,28 @@ export async function fConnect(req: Request, res: Response): Promise<void> {
...
+        data: link,
+    };
+    res.json(callResponse);
+}
+
+export async function fOauth2Complete(req: Request, res: Response): Promise<void> {
+    const call: AppCall = req.body;
+    const context = contextFromRequest(req);
+    const code = call.values.code;
+    if (code === '') {
```
Comment (mickmister) [inline src/restapi/fConnect.ts:48]:
```suggestion
    if (!code) {
```

Comment (mickmister) [reply src/restapi/fConnect.ts:48]: Is this form value marked as required?

File: src/restapi/fIncomingWebhooks.ts [hunk 14]
```diff
@@ -11,31 +10,37 @@ import {newZDClient, newMMClient} from '../clients';
...
-    };
-    const zdClient = await newZDClient(fakeBotContext);
+    const ticketID = values[TriggerFields.TicketIDKey];
+    const channelID = values[TriggerFields.ChannelIDKey];
+
+    const config = await newConfigStore(context).getValues();
+    const zdHost = config.zd_node_host;
+
+    const token = config.zd_oauth_access_token;
+    if (token === '') {
```
Comment (mickmister) [inline src/restapi/fIncomingWebhooks.ts:23]:
```suggestion
    if (!token) {
```

File: src/restapi/fTarget.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    config.storeConfigInfo(cValues);
+    return `Successfully created ${link}`;
+}
+
+function getTargetUrl(context: AppContext): string {
+    const whSecret = context.app.webhook_secret;
+    const pluginName = getManifest().app_id;
+    const whPath = Routes.App.SubscribeIncomingWebhookPath;
+    let url = context.mattermost_site_url;
+    url += `/plugins/com.mattermost.apps/apps/${pluginName}${whPath}?secret=`;
```
Comment (mickmister) [inline src/restapi/fTarget.ts:67]: Can we have a `getAppsPluginURL(siteURL)` function? Also `pluginName` should be `appID` here

File: src/utils/utils.ts [hunk 16]
```diff
@@ -87,3 +105,22 @@ export function getBulletedList(pretext: string, items: string[]): string {
...
+export function isConfigured(context: AppContext): boolean {
+    return Boolean(context.oauth2.client_id && context.oauth2.client_secret);
+}
+
+export function isUserSystemAdmin(context: AppContext): boolean {
+    return Boolean(context.acting_user.roles && context.acting_user.roles.includes(GeneralConstants.SYSTEM_ADMIN_ROLE));
+}
+
+export function isConnected(context: AppContext): boolean {
+    if (context.oauth2.user && context.oauth2.user.access_token && context.oauth2.user.access_token !== '') {
```
Comment (mickmister) [inline src/utils/utils.ts:118]: The check for empty string here is redundant

```suggestion
    if (context.oauth2.user && context.oauth2.user.access_token) {
```

File: src/utils/utils.ts [hunk 17]
```diff
@@ -87,3 +105,22 @@ export function getBulletedList(pretext: string, items: string[]): string {
...
+        return true;
+    }
+    return false;
+}
+
+export function webhookConfigured(config: AppConfigStore): boolean {
+    return Boolean(config.zd_target_id && config.zd_target_id !== '');
```
Comment (mickmister) [inline src/utils/utils.ts:125]:
```suggestion
    return Boolean(config.zd_target_id);
```

**Review by mickmister — COMMENTED** [review]
No issues found, just some comments on checking for falsy strings

**Pull request merged**